packages/widgets/src/dns-hole/controls/component.tsx

```diff
@@ -0,0 +1,170 @@
+  );
+  const [opened, { close, open }] = useDisclosure(false);
+
+  const [data] = clientApi.widget.dnsHole.summary.useSuspenseQuery(
+    {
+      integrationIds,
+    },
+    {
+      refetchOnMount: false,
+      retry: false,
+    },
+  );
+
+  useEffect(() => {
+    const newStatus = data.map((integrationData) => ({
+      integrationId: integrationData.integrationId,
+      enabled: integrationData.summary.status === "enabled",
+    }));
+    setStatus(newStatus);
+  }, [data]);
+
+  const { mutate: enableDns } = clientApi.widget.dnsHole.enable.useMutation({
+    onSuccess: (_, variables) => {
+      setStatus((prevStatus) =>
+        prevStatus.map((item) => (item.integrationId === variables.integrationId ? { ...item, enabled: true } : item)),
+      );
+    },
+  });
+  const { mutate: disableDns } = clientApi.widget.dnsHole.disable.useMutation({
+    onSuccess: (_, variables) => {
+      setStatus((prevStatus) =>
+        prevStatus.map((item) => (item.integrationId === variables.integrationId ? { ...item, enabled: false } : item)),
+      );
+    },
+  });
+  const toggleDns = (integrationId: string) => {
+    const integrationStatus = status.find((item) => item.integrationId === integrationId);
+    if (integrationStatus?.enabled) {
+      disableDns({ integrationId, duration: 0 });
+    } else {
+      enableDns({ integrationId });
+    }
+  };
+
+  const allEnabled = status.every((item) => item.enabled);
+  const allDisabled = status.every((item) => !item.enabled);
+
+  const ControlsCard = (integrationId: string, integrationKind: string) => {
+    const integrationStatus = status.find((item) => item.integrationId === integrationId);
+    const isEnabled = integrationStatus?.enabled ?? false;
+    const integrationDef = integrationKind === "piHole" ? integrationDefs.piHole : integrationDefs.adGuardHome;
+
+    return (
+      <Card key={integrationId} withBorder={true} m="2.5cqmin" p="2.5cqmin" radius="md">
+        <Flex>
+          <Box m="1.5cqmin" p="1.5cqmin">
+            <Image src={integrationDef.iconUrl} width={50} height={50} fit="contain" />
+          </Box>
+          <Flex direction="column" m="1.5cqmin" p="1.5cqmin" gap="1cqmin">
+            <Badge variant="default">{integrationDef.name}</Badge>
+            <Flex direction="row" gap="2cqmin">
+              <UnstyledButton onClick={() => toggleDns(integrationId)}>
+                <Badge variant="dot" color={dnsLightStatus(isEnabled)}>
+                  {isEnabled
+                    ? t("widget.dnsHoleControls.controls.enabled")
+                    : t("widget.dnsHoleControls.controls.disabled")}
+                </Badge>
+              </UnstyledButton>
+              <ActionIcon disabled={!isEnabled} size={20} radius="xl" top="2.67px" variant="default" onClick={open}>
+                <IconClockPause size={20} color="red" />
+              </ActionIcon>
+            </Flex>
+          </Flex>
+        </Flex>
+      </Card>
+    );
+  };
```

Comment:
Avoid declaring components within another component; it will be redefined each render interval.
Move it outside the component (but in the same file) and simply don't export it.